### PR TITLE
pkg/prometheus: add support for web TLS configuration

### LINKED ISSUE #3782: Support web TLS configuration for Prometheus
Hi team I want to config TLS for Prometheus which is available after 2.24.0 of Prometheus and i have sent required files in the form of secrets and i have started the server with those files all is done in logs TLS enabled and shows server is read but the container is not at starting it is still on 1/2 only  and how to enable the web command line flag like --web.config.file 

```
level=info ts=2021-01-08T04:21:18.493Z caller=main.go:364 msg="Starting Prometheus" version="(version=2.24.0, branch=HEAD, revision=02e92236a8bad3503ff5eec3e04ac205a3b8e4fe)"
level=info ts=2021-01-08T04:21:18.493Z caller=main.go:369 build_context="(go=go1.15.6, user=root@d9f90f0b1f76, date=20210106-13:48:37)"
level=info ts=2021-01-08T04:21:18.493Z caller=main.go:370 host_details="(Linux 3.10.0-1127.18.2.el7.x86_64 #1 SMP Mon Jul 20 22:32:16 UTC 2020 x86_64 prometheus-test-0 (none))"
level=info ts=2021-01-08T04:21:18.493Z caller=main.go:371 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2021-01-08T04:21:18.493Z caller=main.go:372 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2021-01-08T04:21:18.495Z caller=web.go:530 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2021-01-08T04:21:18.495Z caller=main.go:738 msg="Starting TSDB ..."
level=info ts=2021-01-08T04:21:18.497Z caller=tls_config.go:223 component=web msg="TLS is enabled." http2=true
level=info ts=2021-01-08T04:21:18.499Z caller=head.go:645 component=tsdb msg="Replaying on-disk memory mappable chunks if any"
level=info ts=2021-01-08T04:21:18.499Z caller=head.go:659 component=tsdb msg="On-disk memory mappable chunks replay completed" duration=6.609µs
level=info ts=2021-01-08T04:21:18.499Z caller=head.go:665 component=tsdb msg="Replaying WAL, this may take a while"
level=info ts=2021-01-08T04:21:18.500Z caller=head.go:717 component=tsdb msg="WAL segment loaded" segment=0 maxSegment=0
level=info ts=2021-01-08T04:21:18.500Z caller=head.go:722 component=tsdb msg="WAL replay completed" checkpoint_replay_duration=18.058µs wal_replay_duration=913.077µs total_replay_duration=960.915µs
level=info ts=2021-01-08T04:21:18.501Z caller=main.go:758 fs_type=XFS_SUPER_MAGIC
level=info ts=2021-01-08T04:21:18.501Z caller=main.go:761 msg="TSDB started"
level=info ts=2021-01-08T04:21:18.501Z caller=main.go:887 msg="Loading configuration file" filename=/etc/prometheus/config_out/prometheus.env.yaml
level=info ts=2021-01-08T04:21:18.501Z caller=main.go:918 msg="Completed loading of configuration file" filename=/etc/prometheus/config_out/prometheus.env.yaml totalDuration=308.169µs remote_storage=2.871µs web_handler=219ns query_engine=857ns scrape=257.746µs scrape_sd=3.033µs notify=2.266µs notify_sd=1.265µs rules=2.479µs
level=info ts=2021-01-08T04:21:18.501Z caller=main.go:710 msg="Server is ready to receive web requests."

```

### PR DESCRIPTION
**Pull request by fpetkovski**
Prometheus 2.24.0 adds support for TLS and basic auth on HTTP endpoints.
This commit enables configuring TLS in the PrometheusSpec.

Fixes #3782

```release-note:feature
added web tls configuration support for prometheus
```

### CODE REVIEW

File: pkg/webconfig/tls_credentials.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+func (a tlsCredentials) getCAMountPath() string {
+	if a.clientCA.ConfigMap != nil || a.clientCA.Secret != nil {
+		return a.assetPath(a.clientCA)
+	}
+
+	return ""
+}
+
+func (a tlsCredentials) assetPath(sel monitoringv1.SecretOrConfigMap) string {
+	return path.Join(a.mountPath, assets.TLSAssetKeyFromSelector("", sel).String())
```
Comment (fpetkovski) [inline pkg/webconfig/tls_credentials.go:82]: Here we are reusing the functionality from the assets package to generate a path inside the container. I am not sure if this is a good idea though, as changing the assets package will affect this one as well. I am open to any suggestions.

File: bundle.yaml [hunk 2]
```diff
@@ -10190,6 +10190,118 @@ spec:
                   pageTitle:
                     description: The prometheus web page title
                     type: string
+                  tlsConfig:
```
Comment (paulfantom) [inline bundle.yaml:10193]: Could you move generated content into separate PR? It will help with the review.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -141,7 +141,7 @@ jobs:
       run: >
         EXCLUDE_ALERTMANAGER_TESTS=${{ matrix.alertmanager }}
         EXCLUDE_PROMETHEUS_TESTS=${{ matrix.prometheus }}
-        EXCLUDE_THANOS_TESTS=${{ matrix.thanosruler }}
+        EXCLUDE_THANOSRULER_TESTS=${{ matrix.thanosruler }}
```
Comment (paulfantom) [inline .github/workflows/ci.yaml:144]: This change is not part of what is described in PR. Could it be moved to a separate commit or separate PR?

File: example/web-tls/secret.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
...NGd6bEpYdG1lSnVMRU81TGlINVdUdjlnMWxBZ01CQUFHalV6QlJNRThHQTFVZEVRUklNRWFDQ25CeWIyMWwKZEdobGRYT0NFbkJ5YjIxbGRHaGxkWE11WkdWbVlYVnNkSUlrY0hKdmJXVjBhR1YxY3k1a1pXWmhkV3gwTG5OMgpZeTVqYkhWemRHVnlMbXh2WTJGc01BMEdDU3FHU0liM0RRRUJDd1VBQTRJQ0FRQ29jSFh6VzVxSU9PVDJJY1gwCnJLaGpvS0hlTWFVOUR3bW1MdG80RTQ4VHlrazhsQ1p3TjhkUXFhYXA4TFczSFdReGZvWWJDSHlSTkRxZCtRdmoKenNCdUZhcUJJZ0ZHcDlsbERhNWtLYlN2cXpSZmdTWWVPVG5keTBEdFlMYTVOU1p4d05VVmRmSXRVY2o2MW1vWgprd01odWhPa21kc3E1R2NKRng4dmhwUWdZN3M2UDVmV3RhbkwwZEZLNEVRbWN6OHpoNjh6TFB0dTYzWStWL1o1CjBhZ0FEL09iazdITUxkSjZjQlNCczJXcnpNZm5CaENHWWVQb3ZneGs2ejA1bThSeU84WFN2S1VwVHhHd0JSUmgKN0dxSXdEWmMvWjNNdEFjUTAyUXRoVzBiQlhwVHQrcS9PaXZOdHBXK0pMT0wrSUVGSXBBKzllWUpWbCt4NnFCMwo2aVRYc3l5aG5waGdFKzR5V0FuODdKcUZ3Ly9ZL1dmbk9oU1hVZFBFZThrZDl3VDNqRzdvd3AvSFNLOFcwU04wCmdRODNQelVZSHU3bDVrOEFIR1llNmd4WkpoUUZDTkVTWXlOalV5Ulk1Mk5oU2QrdENRcHFqdnA3T2kvZjBWWTcKTzl4WjlKeUord2F6Zk03UFl3MXAzNldWcnhqRDYzdGlnVjc5YmVXbXFZZTZuenpHYkJzN3hQaW02djJVdmd5ZApEL2I5dmIzSUVLbDd5UlQvamlOV2RET2RyczVJenVNYjd1T2NXY21FM2RmeXhvWU9tUG1QL2oxdmtCQ3lCdlhMCm0vMlRGclZJV1RPNUVES0EybmY0bmdTakNmUy8vSnQvdk1Ld002dEt0VER5ekFIcWFpYWtyTmdWNkphemNBSi8KcHpMbVFGMTNHU1J4Snd2Z000NGUyUnFlbHc9PQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
```
Comment (paulfantom) [inline example/web-tls/secret.yaml:3]: If we need it as an example, can we set a validity date for sometime in the past (so people will need to generate their own certs)?

File: pkg/apis/monitoring/v1/types.go [hunk 5]
```diff
@@ -503,7 +504,64 @@ type QuerySpec struct {
...
+		return fmt.Errorf("web.tlsConfig.cert must be defined")
+	} else if err := c.Cert.Validate(); err != nil {
+		return err
+	}
+
+	if c.KeySecret == (v1.SecretKeySelector{}) {
+		return fmt.Errorf("web.tlsConfig.key must be defined")
+	}
+
+	return nil
```
Comment (paulfantom) [inline pkg/apis/monitoring/v1/types.go:564]: WDYT about following a pattern from https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/apis/monitoring/v1/types.go#L976-L1034 ?

File: test/framework/resources/prometheus-web-tls-secret.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
...NGd6bEpYdG1lSnVMRU81TGlINVdUdjlnMWxBZ01CQUFHalV6QlJNRThHQTFVZEVRUklNRWFDQ25CeWIyMWwKZEdobGRYT0NFbkJ5YjIxbGRHaGxkWE11WkdWbVlYVnNkSUlrY0hKdmJXVjBhR1YxY3k1a1pXWmhkV3gwTG5OMgpZeTVqYkhWemRHVnlMbXh2WTJGc01BMEdDU3FHU0liM0RRRUJDd1VBQTRJQ0FRQ29jSFh6VzVxSU9PVDJJY1gwCnJLaGpvS0hlTWFVOUR3bW1MdG80RTQ4VHlrazhsQ1p3TjhkUXFhYXA4TFczSFdReGZvWWJDSHlSTkRxZCtRdmoKenNCdUZhcUJJZ0ZHcDlsbERhNWtLYlN2cXpSZmdTWWVPVG5keTBEdFlMYTVOU1p4d05VVmRmSXRVY2o2MW1vWgprd01odWhPa21kc3E1R2NKRng4dmhwUWdZN3M2UDVmV3RhbkwwZEZLNEVRbWN6OHpoNjh6TFB0dTYzWStWL1o1CjBhZ0FEL09iazdITUxkSjZjQlNCczJXcnpNZm5CaENHWWVQb3ZneGs2ejA1bThSeU84WFN2S1VwVHhHd0JSUmgKN0dxSXdEWmMvWjNNdEFjUTAyUXRoVzBiQlhwVHQrcS9PaXZOdHBXK0pMT0wrSUVGSXBBKzllWUpWbCt4NnFCMwo2aVRYc3l5aG5waGdFKzR5V0FuODdKcUZ3Ly9ZL1dmbk9oU1hVZFBFZThrZDl3VDNqRzdvd3AvSFNLOFcwU04wCmdRODNQelVZSHU3bDVrOEFIR1llNmd4WkpoUUZDTkVTWXlOalV5Ulk1Mk5oU2QrdENRcHFqdnA3T2kvZjBWWTcKTzl4WjlKeUord2F6Zk03UFl3MXAzNldWcnhqRDYzdGlnVjc5YmVXbXFZZTZuenpHYkJzN3hQaW02djJVdmd5ZApEL2I5dmIzSUVLbDd5UlQvamlOV2RET2RyczVJenVNYjd1T2NXY21FM2RmeXhvWU9tUG1QL2oxdmtCQ3lCdlhMCm0vMlRGclZJV1RPNUVES0EybmY0bmdTakNmUy8vSnQvdk1Ld002dEt0VER5ekFIcWFpYWtyTmdWNkphemNBSi8KcHpMbVFGMTNHU1J4Snd2Z000NGUyUnFlbHc9PQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
```
Comment (paulfantom) [inline test/framework/resources/prometheus-web-tls-secret.yaml:3]: This certificate should have a longer validity period otherwise tests will start to fail after `May  5 07:25:04 2022 GMT`. Let's increase it to 10 years instead of a year.

**Review by paulfantom — CHANGES_REQUESTED** [review]

File: pkg/apis/monitoring/v1/types.go (hunk 5 above)
Comment (fpetkovski) [reply pkg/apis/monitoring/v1/types.go]: Sounds good, refactored to match the pattern

File: bundle.yaml (hunk 2 above)
Comment (fpetkovski) [reply bundle.yaml]: Should I add the generated assets as a new commit or a new PR? Wouldn't a new PR break atomic changes which are handy in case of rollbacks?

Comment (paulfantom) [reply bundle.yaml]: As a separate commit in the same PR.

Comment (fpetkovski) [reply bundle.yaml:10193]: Roger. Commits should be now split into written and auto-generated code.

File: pkg/webconfig/config.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+var (
+	volumeName = "web-config"
+	configFile = "web-config.yaml"
+)
+
+// Config is the web configuration for a prometheus instance.
+//
+// Config can make a secret which holds the web config contents, as well as
+// volumes and volume mounts for referencing the secret and the
+// necessary TLS credentials
```
Comment (simonpasquier) [inline pkg/webconfig/config.go:35]:
```suggestion
// necessary TLS credentials.
```

**Review by simonpasquier — COMMENTED** [review]
just a few minor comments, very good otherwise!

File: test/e2e/prometheus_test.go [hunk 8]
```diff
@@ -3651,6 +3652,49 @@ func testPromSecurePodMonitor(t *testing.T) {
...
+						Name: "web-tls",
+					},
+					Key: "tls.crt",
+				},
+			},
+		},
+	}
+	if _, err := framework.CreatePrometheusAndWaitUntilReady(ns, prom); err != nil {
+		t.Fatal("Creating prometheus failed: ", err)
+	}
```
[the remark this answers is not in the document]
Comment (fpetkovski) [reply test/e2e/prometheus_test.go:3697]: Good idea, this should be there now

**Review by simonpasquier — APPROVED** [review]
:+1: very nice! It only needs a rebase :)

**Comment by fpetkovski** [comment]
Thanks for the review @simonpasquier. The PR should be rebased now :)
